### ISSUE TITLE
PHPUnit executes using the binary version rather than the one imported through composer

### ISSUE THREAD

**Issue by erichmusick** [MicrosoftDocs/vsts-docs#4011]
I cloned the repo and added the configuration under "Test with phpunit" to my azure-pipelines.yml file, pushed it to DevOps, and ran the pipeline.

Composer installs PHPUnit 4.8.34, while the unit tests execute against PHPUnit 7.5.2. This results in different output when running in DevOps than when executing `./phpunit`. 

Composer output:
- Installing phpunit/phpunit (4.8.34)

"Run tests with phpunit" output:
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/f4ca2c23-5750-4699-a2ff-3d2db7176d53.sh
PHPUnit 7.5.8 by Sebastian Bergmann and contributors.

I think that the script needs to invoke the phpunit *script* rather than the executable. So, I changed the script to `./phpunit`, but then DevOps reports an error:

/home/vsts/work/_temp/0455c56f-5067-4ee2-b598-32a968b844ea.sh: line 1: ./phpunit: Permission denied
##[error]Bash exited with code '126'.

So, not sure how to proceed here. But something isn't quite right.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 4d21b6ef-2440-f2b7-7f38-465ade7c14c6
* Version Independent ID: 9b2ebcef-924d-6850-9c27-4f14a2b45b53
* Content: [Build and test PHP apps - Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/languages/php?view=azure-devops)
* Content Source: [docs/pipelines/languages/php.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/pipelines/languages/php.md)
* Product: **devops**
* Technology: **devops-cicd**
* GitHub Login: @thomps23
* Microsoft Alias: **dastahel**

**Comment by erichmusick** [comment]
Here's what I had to do to get the pipeline to run the `phpunit` **script** in the repo, rather than the `phpunit` **binary**:

```yml
- script: sudo chmod 0755 ./phpunit
  displayName: 'Make phpunit script runnable'

- script: ./phpunit
  displayName: 'Run tests with phpunit'
  continueOnError: true
```

**Comment by kmdeva** [comment]
Hi,
    If I use .phpunit, I get the below error 
Task         : Command Line
Description  : Run a command line script using cmd.exe on Windows and bash on macOS and Linux.
Version      : 2.148.0
Author       : Microsoft Corporation
Help         : https://go.microsoft.com/fwlink/?LinkID=613735
==============================================================================
Generating script.
Script contents:
sudo chmod 0755 ./phpunit
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/514ac076-ce08-4728-867f-abb359680e43.sh
chmod: cannot access './phpunit': No such file or directory
##[error]Bash exited with code '1'.
If I use only phpunit , I get error code 2  .. Please suggest

**Comment by erichmusick** [comment]
@kmdeva - Are you using the [PHP Pipelines Repo](https://github.com/MicrosoftDocs/pipelines-php) or your own repo? The sample repo has a `phpunit` [script](https://github.com/MicrosoftDocs/pipelines-php/blob/master/phpunit) which this command is trying to execute. This error suggests that the file is missing from the repo.

**Comment by kmdeva** [comment]
Hello Erich,
        Thank you for the reply. Yes we got this command in the repo.We are using Aquia BLT repo for our pipeline and we use Azure pipeline to trigger this. As per this document from Microsoft we used these commands and now I am facing the below error ..I have tried all the combination of argument and options
==============================================================================
Task         : Command Line
Description  : Run a command line script using cmd.exe on Windows and bash on macOS and Linux.
Version      : 2.148.0
Author       : Microsoft Corporation
Help         : https://go.microsoft.com/fwlink/?LinkID=613735
==============================================================================
Generating script.
Script contents:
phpunit
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/5c7fe1ce-bc7c-4792-9f5e-1ac5595b73db.sh
PHPUnit 7.5.11 by Sebastian Bergmann and contributors.

Usage: phpunit [options] UnitTest [UnitTest.php]
       phpunit [options] <directory>

Code Coverage Options:

  --coverage-clover <file>    Generate code coverage report in Clover XML format
  --coverage-crap4j <file>    Generate code coverage report in Crap4J XML format
  --coverage-html <dir>       Generate code coverage report in HTML format
  --coverage-php <file>       Export PHP_CodeCoverage object to file
  --coverage-text=<file>      Generate code coverage report in text format
                              Default: Standard output
  --coverage-xml <dir>        Generate code coverage report in PHPUnit XML format
  --whitelist <dir>           Whitelist <dir> for code coverage analysis
  --disable-coverage-ignore   Disable annotations for ignoring code coverage
  --no-coverage               Ignore code coverage configuration
  --dump-xdebug-filter <file> Generate script to set Xdebug code coverage filter

Logging Options:

  --log-junit <file>          Log test execution in JUnit XML format to file
  --log-teamcity <file>       Log test execution in TeamCity format to file
  --testdox-html <file>       Write agile documentation in HTML format to file
  --testdox-text <file>       Write agile documentation in Text format to file
  --testdox-xml <file>        Write agile documentation in XML format to file
  --reverse-list              Print defects in reverse order

Test Selection Options:

  --filter <pattern>          Filter which tests to run
  --testsuite <name,...>      Filter which testsuite to run
  --group ...                 Only runs tests from the specified group(s)
  --exclude-group ...         Exclude tests from the specified group(s)
  --list-groups               List available test groups
  --list-suites               List available test suites
  --list-tests                List available tests
  --list-tests-xml <file>     List available tests in XML format
  --test-suffix ...           Only search for test in files with specified
                              suffix(es). Default: Test.php,.phpt

Test Execution Options:

  --dont-report-useless-tests Do not report tests that do not test anything
  --strict-coverage           Be strict about @covers annotation usage
  --strict-global-state       Be strict about changes to global state
  --disallow-test-output      Be strict about output during tests
  --disallow-resource-usage   Be strict about resource usage during small tests
  --enforce-time-limit        Enforce time limit based on test size
  --default-time-limit=<sec>  Timeout in seconds for tests without @small, @medium or @large
  --disallow-todo-tests       Disallow @todo-annotated tests

  --process-isolation         Run each test in a separate PHP process
  --globals-backup            Backup and restore $GLOBALS for each test
  --static-backup             Backup and restore static attributes for each test

  --colors=<flag>             Use colors in output ("never", "auto" or "always")
  --columns <n>               Number of columns to use for progress output
  --columns max               Use maximum number of columns for progress output
  --stderr                    Write to STDERR instead of STDOUT
  --stop-on-defect            Stop execution upon first not-passed test
  --stop-on-error             Stop execution upon first error
  --stop-on-failure           Stop execution upon first error or failure
  --stop-on-warning           Stop execution upon first warning
  --stop-on-risky             Stop execution upon first risky test
  --stop-on-skipped           Stop execution upon first skipped test
  --stop-on-incomplete        Stop execution upon first incomplete test
  --fail-on-warning           Treat tests with warnings as failures
  --fail-on-risky             Treat risky tests as failures
  -v|--verbose                Output more verbose information
  --debug                     Display debugging information

  --loader <loader>           TestSuiteLoader implementation to use
  --repeat <times>            Runs the test(s) repeatedly
  --teamcity                  Report test execution progress in TeamCity format
  --testdox                   Report test execution progress in TestDox format
  --testdox-group             Only include tests from the specified group(s)
  --testdox-exclude-group     Exclude tests from the specified group(s)
  --printer <printer>         TestListener implementation to use

  --resolve-dependencies      Resolve dependencies between tests
  --order-by=<order>          Run tests in order: default|reverse|random|defects|depends
  --random-order-seed=<N>     Use a specific random seed <N> for random order
  --cache-result              Write run result to cache to enable ordering tests defects-first

Configuration Options:

  --prepend <file>            A PHP script that is included as early as possible
  --bootstrap <file>          A PHP script that is included before the tests run
  -c|--configuration <file>   Read configuration from XML file
  --no-configuration          Ignore default configuration file (phpunit.xml)
  --no-logging                Ignore logging configuration
  --no-extensions             Do not load PHPUnit extensions
  --include-path <path(s)>    Prepend PHP's include_path with given path(s)
  -d key[=value]              Sets a php.ini value
  --generate-configuration    Generate configuration file with suggested settings
  --cache-result-file=<file>  Specify result cache path and filename

Miscellaneous Options:

  -h|--help                   Prints this usage information
  --version                   Prints the version and exits
  --atleast-version <min>     Checks that version is greater than min and exits
  --check-version             Check whether PHPUnit is the latest version
##[error]Bash exited with code '2'.

**Comment by kmdeva** [comment]
Referred this link 
https://docs.microsoft.com/en-us/azure/devops/pipelines/languages/php?view=azure-devops

**Comment by erichmusick** [comment]
@kmdeva - The issue you're describing sounds like a different root cause than what I reported. The problem I've found is specifically with the sample PHP Pipelines repo linked from the post you shared, and the ambiguity around needing to set permissions on the `phpunit` script.

If you're using the vanilla [Acquia BLT repo](https://github.com/acquia/blt), I see two possible problems:
1. When you use `./phpunit`, this fails because there's no script called `phpunit` in the root of the repo. Aside: I'm not sure if this is a convention in PHP projects, but it's confusing that this _script_ has no file extension; I think the whole situation would be clearer if it were called `phpunit.sh`, to clearly differentiate from the `phpunit` binary.
2. When you use `phpunit`, this fails because there are no command line arguments and no PHPUnit configuration (i.e. `phpunit.xml`) to instruct the executable which tests to run. According to the PHPUnit docs, I think it should try to read `phpunit.xml.dist` from the current working directory, so I might be mistaken on this point.

I'd suggest trying each of the following to see if one works:
1. Change the command to `phpunit -c phpunit.xml.dist`
2. Rename `phpunit.xml.dist` to `phpunit.xml`, and use `phpunit`
3. Add a `phpunit` script like [the one](https://github.com/MicrosoftDocs/pipelines-php/blob/master/phpunit) in the sample [Azure Pipelines PHP Repo](https://github.com/MicrosoftDocs/pipelines-php), and use `./phpunit`, with an additional step to set permissions in the script (see the example I shared in my first reply to the original post)

**Comment by jake-harris** [comment]
Hi @erichmusick 

I think the discrepancy in version may be due to phpunit 7+ being installed by some other process "globally. To use your project-specific composer binary, maybe try using the binary located at `/vendor/bin/composer`. I haven't tested this, so forgive me if it's a snipe hunt.

**Comment by WilliamAntonRohm** [comment]
@erichmusick -- Erich, you may find answers here:
- [Azure DevOps Support Bot](https://azuredevopsvirtualagent.azurewebsites.net/) 
- [Azure DevOps on Stack Overflow](https://stackoverflow.com/questions/tagged/azure-devops)

I am closing this issue now.  You are welcome to @ mention me for any followup.
We hope to hear from you again.

My apologies for the delay in responding.

**Comment by erichmusick** [comment]
Hey @WilliamAntonRohm ... sorry for the late follow up. I still see two issues. I would love to be able to clone the sample repo and copy/paste the steps from the guide and run the pipeline successfully. As-is, this doesn't work.

I was able to get this all to work with a bit of wrangling. This required two changes:

1. In the pipelines-php repo, we need to commit the `phpunit` script with execute permissions (e.g. 755). I'll send a PR for this. (As an aside, I think this whole matter would have been less confusing had this file been named phpunit.sh, thus avoiding the conflict with the `phpunit` binary)

2. The instructions should reference either the phpunit binary in vendor folder, as @jake-harris mentioned, OR the phpunit script in the repo. I was able to do the latter by changing the step in the guide to `./phpunit` rather than `phpunit`:

```
- script: ./phpunit
  displayName: 'Run tests with phpunit'
```

I'll look at sending a PR for this, too.
